hail/src/test/scala/is/hail/expr/ir/IRSuite.scala

```diff
@@ -1119,6 +1119,29 @@ class IRSuite extends SparkSuite {
       6L)
   }
 
+  @Test def testArrayAggContexts() {
+    implicit val execStrats = Set(ExecStrategy.JvmCompile)
+
+    val ir = Let(
+      "x",
+      In(0, TInt64()) * In(0, TInt64()),
+      ArrayAgg(
+        ArrayRange(I32(0), I32(10), I32(1)),
+        "elt",
+        AggLet("y",
+          Cast(Ref("x", TInt64()) * Ref("x", TInt64()) * Cast(Ref("elt", TInt32()), TInt64()), TInt32()),
+          Cast(Ref("x", TInt64()), TInt32()) +
+            ApplyAggOp(FastIndexedSeq(), None, FastIndexedSeq(
+              Ref("elt", TInt32()) + Cast(Ref("x", TInt64()), TInt32()) + Ref("y", TInt32()) + Ref("y", TInt32())),
+              AggSignature(Max(), FastIndexedSeq(), None, FastIndexedSeq(TInt32()))),
+          isScan = false
+        )
+      )
+    )
+
+    assertEvalsTo(ir, FastIndexedSeq(1L -> TInt64()), 9 * 3 + 1 + 1)
```

Comment:
This is a hard test case to read, but it seems like the extra +1 is wrong, no?  In(0) = 1, so x = In(0)*In(0) = 1 and y = elt.  Thus the result should be 9 + 1 + 9 + 9 = 9 * 3 + 1.

Reply:
I'm rewriting it as an agg collect, it's way easier to understand

Reply:
(and reason about)

Reply:
ok, commit is up

Reply:
Wait, was my reasoning of the previous one wrong?  It was passing and I'm still surprised.  I'd like to understand before we move on.

Reply:
OK, this one I buy, but I'd like to know what was up with the previous one.

Reply:
there's an `x + ` both inside the ApplyAggOp and outside it

Reply:
x is 1, y is elt. 

Max (elt + y + y + x) + x

Reply:
Ah, I just missed the last x which was above.  Thanks for clarifying!